src/signals.c

```diff
@@ -2009,3 +2007,113 @@ void widget_file_monitor_try_create(variable *var, gchar *filename)
 	}
 }
 #endif
+
+/***********************************************************************
+ *                                                                     *
+ ***********************************************************************/
+
+void on_any_widget_file_activated_event(GtkWidget *widget, AttributeSet *Attr)
+{
+#ifdef DEBUG_TRANSITS
+	fprintf(stderr, "%s(): Entering.\n", __func__);
+#endif
+
+	widget_signal_executor(widget, Attr, "file-activated");
+
+#ifdef DEBUG_TRANSITS
+	fprintf(stderr, "%s(): Exiting.\n", __func__);
+#endif
+}
+
+/***********************************************************************
+ *                                                                     *
+ ***********************************************************************/
+
+void on_any_widget_current_folder_changed_event(GtkWidget *widget, AttributeSet *Attr)
+{
+#ifdef DEBUG_TRANSITS
+	fprintf(stderr, "%s(): Entering.\n", __func__);
+#endif
+
+	widget_signal_executor(widget, Attr, "current-folder-changed");
+
+#ifdef DEBUG_TRANSITS
+	fprintf(stderr, "%s(): Exiting.\n", __func__);
+#endif
+}
+
+/***********************************************************************
+ * Chooser                                                             *
+ ***********************************************************************/
+
+/* step 2022: Work-around for GTK+-3 GtkFileChooser duplicate events issue.
+ *
+ * The GTK3 file chooser widget triggers duplicate "selection-changed" and
+ * "update-preview" events when fs-filters(-mime) is used to filter the file
+ * list. See demo script "examples/chooser/chooser_event_counters".  Duplicates
+ * do not occur with the GTK2 file chooser.  I have implemented a work-around
+ * by defining a dedicated signal handling function for each of these two
+ * signals.  The handler calls widget_signal_executor only if the current
+ * widget data is non-empty and differs from the previous call, otherwise the
+ * executor isn't called.  Thus, a repeated sequence of the same file path hits
+ * execution only once, and never if the file path is empty.  This logic can
+ * only work because the signal executor handles all actions defined for a
+ * widget as a whole unit. It would not work with the older chooser
+ * implementation, which handled each <action when> independently of other
+ * actions.  Without this work-around, the demo script was showing literally
+ * thousands of duplicate or empty events when changing fs-filters in
+ * directory holding a large photo collection. The GTK2 file chooser benefits
+ * from this work-around too, because it is known to trigger signals when the
+ * file path is empty. */
+
+void on_chooser_widget_selection_changed_event(GtkWidget *widget, AttributeSet *Attr)
+{
+	variable        *var   = NULL;
+	static gchar    *prev  = NULL;
```

Comment:
How big is it? Maybe it's a better idea to store a hash here?

Reply:
`var` and `prev` are pointer-size, I wouldn't use a hash.

Reply:
But if the selection is a large piece of text, you have to hold the previous selection in RAM. If you hash it, say, with CRC32, you only need to store 4 bytes regardless of the previous selection size.

Reply:
It's a pathname.